src/buffer/out/textBuffer.cpp

```diff
@@ -2353,3 +2354,86 @@ void TextBuffer::CopyHyperlinkMaps(const TextBuffer& other)
     _hyperlinkMap = other._hyperlinkMap;
     _hyperlinkCustomIdMap = other._hyperlinkCustomIdMap;
 }
+
+// Method Description:
+// - Adds a regex pattern we should search for
+// - The searching does not happen here, we only search when asked to by TerminalCore
+// Arguments:
+// - The regex pattern
+// Return value:
+// - An ID that the caller should associate with the given pattern
+const size_t TextBuffer::AddPatternRecognizer(const std::string_view regexString)
+{
+    ++_currentPatternId;
+    _IdsAndPatterns.emplace(std::make_pair(_currentPatternId, regexString));
+    return _currentPatternId;
+}
+
+// Method Description:
+// - Finds patterns within the requested region of the text buffer
+// Arguments:
+// - The firstRow to start searching from
+// - The lastRow to search
+// Return value:
+// - A vector containing 3-element tuples where
+//      tuple<0>: ID of the pattern
+//      tuple<1>: start coordinate of the pattern
+//      tuple<2>: end coordinate of the pattern
+const std::vector<std::tuple<size_t, COORD, COORD>> TextBuffer::UpdatePatterns(const size_t firstRow, const size_t lastRow) const

```

Comment:
You might consider using til::point, til::rectangle to represent the region that should be scanned. Same goes for the region covering the found pattern. This feels like a similar mechanism to the both the Selection and Find behaviors we already have. And while those two don't super share an implementation (even though they maybe should), we should likely try to avoid making a 3rd completely different way of talking about "digging through the buffer in this region to find (or not) a matching region"

Reply:
Using interval trees for faster lookup/insertion